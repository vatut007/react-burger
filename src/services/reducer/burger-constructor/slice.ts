import { PayloadAction, createSlice, nanoid } from "@reduxjs/toolkit";
import {
  IngredientConstructorList,
  type Ingredient
} from "../../../types/ingredient";

export const burgerConstructorSlice = createSlice({
  name: "burgerConstructor",
  reducerPath: "burgerConstructor",
  initialState: {
    bun: null as Ingredient | null,
    ingredients: [] as IngredientConstructorList,
    ingredientCount: {} as Record<string,number>
  },
  reducers: {
    addIngredient: {
      reducer(
        state,
        action: PayloadAction<{ cartItemId: string; ingredient: Ingredient }>,
      ) {
        if (action.payload.ingredient.type === "bun") {
          state.bun = action.payload.ingredient;
        } else {
          state.ingredients.push({
            ...action.payload.ingredient,
            cart_item_id: action.payload.cartItemId,
          });
          if (action.payload.ingredient._id in state.ingredientCount){
            state.ingredientCount[action.payload.ingredient._id]++
          } else {
            state.ingredientCount[action.payload.ingredient._id]=1
          }
        }
      },
      prepare({ ingredient }: { ingredient: Ingredient }) {
        const cartItemId = nanoid();
        return { payload: { cartItemId, ingredient } };
      },
    },
    removeIngredient(state, action: PayloadAction<{ cartItemId: string, ingeredient_id: string }>) {
      state.ingredients = state.ingredients.filter(
        (item) => item.cart_item_id !== action.payload.cartItemId,
      );
      state.ingredientCount[action.payload.ingeredient_id]--
    },
    moveIngredient(
      state,
      action: PayloadAction<{ oldIndex: string; newIndex: number }>,
    ) {
      const index = state.ingredients.findIndex(
        (item) => item.cart_item_id === action.payload.oldIndex,
      );
      const [ingredient] = state.ingredients.splice(index, 1);
      state.ingredients.splice(action.payload.newIndex, 0, ingredient);
    },
  },
});
