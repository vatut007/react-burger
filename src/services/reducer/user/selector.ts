import type { RootState } from "../root-reducer";

export const selectSelectedUser = (state: RootState) => state.user.user;
export const selectSelectedAccessToken = (state: RootState) =>
  state.user.accessToken;
export const selectSelectedRefreshToken = (state: RootState) =>
  state.user.refreshToken;
