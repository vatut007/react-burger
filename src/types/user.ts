export type User = {
  email: string;
  name: string;
};

export type ResponseGetUser = {
  success: boolean;
  user: User;
};
