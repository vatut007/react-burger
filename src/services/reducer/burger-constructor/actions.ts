import { burgerConstructorSlice } from "./slice";

export const { addIngredient, removeIngredient, moveIngredient } =
  burgerConstructorSlice.actions;
