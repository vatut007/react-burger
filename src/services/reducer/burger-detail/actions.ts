import { ingredientDetailSlice } from "./slice";

export const { selectIngredient } = ingredientDetailSlice.actions;
