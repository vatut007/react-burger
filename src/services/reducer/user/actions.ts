import { userSlice } from "./slice";

export const { addAccessToken, addUser, addRefreshToken, clearUser } =
  userSlice.actions;
