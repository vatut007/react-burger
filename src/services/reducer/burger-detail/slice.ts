import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { Ingredient } from "../../../types/ingredient";

export const ingredientDetailSlice = createSlice({
    name: 'ingedientDetail',
    reducerPath: 'ingedientDetail',
    initialState: {ingredient:null as Ingredient | null},
    reducers:{
        selectIngredient(state, action: PayloadAction<{ ingredient: Ingredient }>) {
            state.ingredient = action.payload.ingredient
        },
    }
})