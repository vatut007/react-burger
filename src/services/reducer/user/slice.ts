import { PayloadAction, createSlice } from "@reduxjs/toolkit";
import { User } from "../../../types/user";

export const userSlice = createSlice({
  name: "user",
  reducerPath: "user",
  initialState: {
    user: null as User | null,
    accessToken: null as string | null,
    refreshToken: null as string | null,
  },
  reducers: {
    addUser(state, action: PayloadAction<User>) {
      state.user = {
        email: action.payload.email,
        name: action.payload.name,
      };
    },
    addAccessToken(state, action: PayloadAction<{ accessToken: string }>) {
      state.accessToken = action.payload.accessToken;
    },
    addRefreshToken(state, action: PayloadAction<{ refreshToken: string }>) {
      state.refreshToken = action.payload.refreshToken;
    },
    clearUser(state) {
      (state.accessToken = null),
        (state.refreshToken = null),
        (state.user = null);
    },
  },
});
