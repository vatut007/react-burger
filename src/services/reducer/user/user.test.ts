import { User } from "../../../types/user";
import { addAccessToken, addRefreshToken, addUser, clearUser } from "./actions";
import { userSlice } from "./slice";
import { afterEach, beforeEach, expect, test, vi } from "vitest";

const { reducer, getInitialState } = userSlice;

const getUser = (): User => ({
  email: "test@test.ru",
  name: "Test",
});

const getStateUserAdd = () => reducer(getInitialState(), addUser(getUser()))

test("add user", () => {
  expect(reducer(getInitialState(), addUser(getUser()))).toMatchSnapshot();
});

test("add Access Token",
  () => {
    expect(
      reducer(getInitialState(), addAccessToken({ accessToken: "bearer" })),
    ).toMatchSnapshot();
  });

test("add Refresh Token",
  () => {
    expect(
      reducer(getInitialState(), addRefreshToken({ refreshToken: 'refresh'})),
    ).toMatchSnapshot();
  });

test("clear User", ()=>{
    expect(
        reducer(getStateUserAdd(), clearUser())
    ).toMatchSnapshot()
})
