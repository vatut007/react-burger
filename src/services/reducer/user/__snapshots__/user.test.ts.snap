// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`add Access Token 1`] = `
{
  "accessToken": "bearer",
  "refreshToken": null,
  "user": null,
}
`;

exports[`add Refresh Token 1`] = `
{
  "accessToken": null,
  "refreshToken": "refresh",
  "user": null,
}
`;

exports[`add user 1`] = `
{
  "accessToken": null,
  "refreshToken": null,
  "user": {
    "email": "test@test.ru",
    "name": "Test",
  },
}
`;

exports[`clear User 1`] = `
{
  "accessToken": null,
  "refreshToken": null,
  "user": null,
}
`;
